R/baselinenowcast.R

```diff
@@ -152,3 +153,258 @@ baselinenowcast.reporting_triangle <- function(
+    pooled_df <- .combine_triangle_dfs(
+      data = data_clean,
+      strata_cols = strata_cols
+    )
+    pooled_triangle <- as_reporting_triangle(pooled_df,
+      max_delay = max_delay
+    )
+    # Get the training volume for all reporting triangles
+    tv <- allocate_reference_times(
+      reporting_triangle = pooled_triangle$reporting_triangle_matrix,
+      scale_factor = scale_factor,
+      prop_delay = prop_delay
+    )
+    if ("delay" %in% strata_sharing) {
+      # Estimate delay once on pooled data
+      delay_pmf <- estimate_delay(
+        reporting_triangle = pooled_triangle$reporting_triangle_matrix,
+        n = tv$n_history_delay
+      )
+    }
+    if ("uncertainty" %in% strata_sharing) {
+      # Estimate uncertainty once on pooled data
+      trunc_rep_tris <- truncate_triangles(
+        reporting_triangle = pooled_triangle$reporting_triangle_matrix,
+        n = tv$n_retrospective_nowcasts
+      )
+      retro_rep_tris <- construct_triangles(
+        trunc_rep_tris,
+        structure = pooled_triangle$structure
+      )
+      retro_pt_nowcasts <- fill_triangles(retro_rep_tris,
+        n = tv$n_history_delay
+      )
+      uncertainty_params <- estimate_uncertainty(
+        point_nowcast_matrices = retro_pt_nowcasts,
+        truncated_reporting_triangles = trunc_rep_tris,
+        retro_reporting_triangles = retro_rep_tris,
+        n = tv$n_retrospective_nowcasts,
+        uncertainty_model = uncertainty_model,
+        ...
+      )
+    }
+  }
+
+  # Nowcast
+  combined_result <- imap(
+    list_of_rep_tris,
+    ~ .nowcast_from_rep_tris(
+      rep_tri = .x,
+      name = .y,
+      strata_cols = strata_cols,
+      scale_factor = scale_factor,
+      prop_delay = prop_delay,
+      output_type = output_type,
+      draws = draws,
+      uncertainty_model =
+        uncertainty_model,
+      uncertainty_sampler =
+        uncertainty_sampler,
+      delay_pmf = delay_pmf,
+      uncertainty_params =
+        uncertainty_params
+    )
+  ) |> list_rbind()
+  return(combined_result)
+}
+
+
+#' Produce a nowcast from a named reporting triangle object
+#'
+#' @param rep_tri Object of reporting triangle class
+#' @param name Name of the reporting triangle class object
+#' @inheritParams baselinenowcast
+#' @inheritParams baselinenowcast.data.frame
+#' @inheritParams baselinenowcast.reporting_triangle
+#' @inheritParams as_reporting_triangle.data.frame
+#' @inheritParams estimate_uncertainty
+#' @inheritParams sample_nowcast
+#' @inheritParams allocate_reference_times
+#' @keywords internal
+#' @returns Data.frame of nowcasts with columns obtained from the name of the
+#'   original list and the nowcast unit.
+.nowcast_from_rep_tris <- function(
+    rep_tri,
+    name,
+    strata_cols,
+    scale_factor,
+    prop_delay,
+    output_type,
+    draws,
+    uncertainty_model,
+    uncertainty_sampler,
+    delay_pmf,
+    uncertainty_params,
+    ref_time_aggregator = identity,
```

Comment:
Do you need these to have defauls? Seems tedious in terms of maintenance?

Reply:
Note -- I had to add `uncertainty_model` as an input argument to `estimate_uncertainty_retro()`. This is because we cannot pass with `...` both `uncertainty_model` and the args to `estimate_uncertainty` of `ref_time_aggregator()` and `delay_aggregator`.

This is an artifact of the fact that `ref_time_aggregator` and `delay_aggregator` are needed for both uncertainty estimation and sampling, whereas `uncertainty_model` is only needed for estimation and `uncertainty_sampler` is only needed for sampling. Therefore, we can't just lump them all into one category with the `....` as we need to specify what goes where.